src/common/QueryHandler.ts

```diff
@@ -14,62 +14,137 @@
     element: Element | Document,
     selector: string
   ) => Element[] | NodeListOf<Element>;
 }
 
-const _customQueryHandlers = new Map<string, QueryHandler>();
+function mkQueryHandler(handler: CustomQueryHandler): QueryHandler {
+  const internalHandler: QueryHandler = {};
+
+  if (handler.queryOne) {
+    internalHandler.queryOne = async (element, selector) => {
+      const jsHandle = await element.evaluateHandle(handler.queryOne, selector);
+      const elementHandle = jsHandle.asElement();
+      if (elementHandle) return elementHandle;
+      await jsHandle.dispose();
+      return null;
+    };
+    internalHandler.waitFor = (
+      domWorld: DOMWorld,
+      selector: string,
+      options: WaitForSelectorOptions
+    ) => domWorld.waitForSelectorInPage(handler.queryOne, selector, options);
+  }
+
+  if (handler.queryAll) {
+    internalHandler.queryAll = async (element, selector) => {
+      const jsHandle = await element.evaluateHandle(handler.queryAll, selector);
+      const properties = await jsHandle.getProperties();
+      await jsHandle.dispose();
+      const result = [];
+      for (const property of properties.values()) {
+        const elementHandle = property.asElement();
+        if (elementHandle) result.push(elementHandle);
+      }
+      return result;
+    };
+    internalHandler.queryAllArray = async (element, selector) => {
+      const resultHandle = await element.evaluateHandle(
+        handler.queryAll,
+        selector
+      );
+      const arrayHandle = await resultHandle.evaluateHandle(
+        (res: Element[] | NodeListOf<Element>) => Array.from(res)
+      );
+      return arrayHandle;
+    };
+  }
+
+  return internalHandler;
+}
+
+const _queryHandlers = new Map<string, QueryHandler>();
+const _defaultHandler = mkQueryHandler({
+  queryOne: (element: Element, selector: string) =>
+    element.querySelector(selector),
+  queryAll: (element: Element, selector: string) =>
+    element.querySelectorAll(selector),
+});
 
 export function registerCustomQueryHandler(
   name: string,
-  handler: QueryHandler
+  handler: CustomQueryHandler
 ): void {
-  if (_customQueryHandlers.get(name))
+  if (_queryHandlers.get(name))
     throw new Error(`A custom query handler named "${name}" already exists`);
 
   const isValidName = /^[a-zA-Z]+$/.test(name);
   if (!isValidName)
     throw new Error(`Custom query handler names may only contain [a-zA-Z]`);
 
-  _customQueryHandlers.set(name, handler);
+  const internalHandler = mkQueryHandler(handler);
+
+  _queryHandlers.set(name, internalHandler);
 }
 
 /**
  * @param {string} name
  */
 export function unregisterCustomQueryHandler(name: string): void {
-  _customQueryHandlers.delete(name);
+  if (_queryHandlers.has(name)) {
+    _queryHandlers.delete(name);
+  }
 }
 
-export function customQueryHandlers(): Map<string, QueryHandler> {
-  return _customQueryHandlers;
+export function customQueryHandlerNames(): string[] {
+  return Array.from(_queryHandlers.keys());
```

Comment:
This is debatable, but I weakly prefer array spread syntax here:

```js
[..._queryHandlers.keys()]
```

WDYT?

Reply:
I keep forgetting that this syntax exist. We'll do that instead :)